source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown

```diff
@@ -0,0 +1,127 @@
+---
+layout: post
+title: "ESP8266 and MicroPython - Part 2"
+description: "Export, process, and visualize data stored by Home Assistant."
+date: 2016-07-25 06:00:00 +0200
+date_formatted: "July 25, 2016"
+author: Fabian Affolter
+comments: true
+categories: How-To
+og_image: /images/blog/2016-07-micropython/social.png
+---
+
+<img src='/images/blog/2016-07-micropython/micropython.png' style='clear: right; border:none; box-shadow: none; float: right; margin-bottom: 12px;' width='200' />
+So, part 1 of [ESP8266 and MicroPython]() was pretty lame, right? Instead of get information out of Home Assistant we are going a step forward. Some of the buzz words are "Sensor", "MQTT", and "Button". 
+
+<!--more-->
+
+Beside HTTP POST requests, MQTT is the quickest way (from the author's point of view) to publish information with DIY devices. 
+
+You have to make a decision: Do you want to pull or to poll? For slowly changing values like temperature it's perfectly fine to wait a couple of seconds to retrieve the value. If it's a motion detector the state change should be available instantly. This means the sensor must take initiative. 
+
+An example for pulling is [aREST](/components/sensor.arest/). This is a great way to work with the ESP8266 based units and the Ardunio IDE. 
+
+Let's do a HTTP POST first with MicroPython. The `curl` command should a new value `"value=23"` as POST to an API endpoint called `/sensor2`.
+
+```bash
+$ curl -X POST -d "value=23" http://192.168.1.18:5000/sensor2
+```
+
+The conversion of the `curl` command to use with `urequests` is very simple. Basically it's a one-liner but splitted to be easier to read.
+ 
+```python
+>>> import urequests
+>>> url = 'http://192.168.1.18:5000/sensor2'
+>>> headers = {'content-type': 'application/json'}
+>>> data = '{"value": 23}'
+>>> resp = urequests.post(url, data=data, headers=headers)
+>>> print(resp.json())
+{'name': 'Sensor2', 'value': 23}
+>>> resp1 = urequests.get(url, headers=headers)
+>>> print(resp1.json())
+{'name': 'Sensor2', 'value': 23}
+```
+
+I think that with this piece of information you should be able to use the [Home Assistant RESTful API](/developers/rest_api/).
+
+Now, MQTT. You can find simple examples for publishing and subscribing for MQTT in the [MicroPython](https://github.com/micropython/micropython-lib) library overview in the section for [umqtt](https://github.com/micropython/micropython-lib/tree/master/umqtt.simple). 
+
+[@davea](https://github.com/davea) created [sonoff-mqtt](https://github.com/davea/sonoff-mqtt). This code will work on ESP8622 based devices too. The example below is adopted from the work of [@davea](https://github.com/davea) as we don't want to re-invent the wheel.
```

Comment:
I don't think that sonoff is/was a stretch goal of the campaign. @davea created it around umqtt which is/was a stretch goal.